### PR TITLE
deprecate get_all_items_as_dict and get_all_items, add get_items_as_dicts

### PR DESCRIPTION
**Pull request by philvarner**
**Related Issue(s):** #187

**Description:**

- deprecate get_all_items_as_dict and get_all_items, as these can retrieve the entire catalog without the user really knowing
- add get_items_as_dicts to allow for retrieval of unmarshalled items 

**PR Checklist:**

- [X] Code is formatted
- [X] Tests pass
- [X] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac-api-client/blob/main/CHANGELOG.md)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,15 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 ## [Unreleased] - TBD
 
+### Deprecated
+
+- Item Search methods `get_all_items` and `get_all_items_as_dict` are now deprecated.
+  These have been deprecated because they have the potential to perform a large number
+  of requests to the server and instantiate a large number of objects in memory.
+  To a user, this is only visible as a large delay in the method call and/or the
+  exhaustion of all available memory. The iterator methods `get_items` or
+  `get_itemcollections` should be used instead.
```
Comment (gadomski) [inline CHANGELOG.md:16]: Do we want to specify when the methods will be removed, or leave that unbounded?

File: pystac_client/item_search.py [hunk 2]
```diff
@@ -591,14 +594,40 @@ def get_items(self) -> Iterator[Item]:
...
+                    return
+
     @lru_cache(1)
-    def get_all_items_as_dict(self) -> Dict:
-        """Convenience method that gets all items from all pages, up to self._max_items,
-         and returns an array of dictionaries
+    def get_all_items_as_dict(self) -> Dict[str, Any]:
+        """DEPRECATED. Use get_items or get_itemcollections instead.
+            Convenience method that gets all items from all pages, up to
+            self._max_items, and returns an array of dictionaries.
```
Comment (gadomski) [inline pystac_client/item_search.py:621]: This isn't introduced in this PR, but I don't love referencing a "private" attribute in the documentation. `_max_items` should either be made public or exposed via a property or method.

**Review by gadomski — CHANGES_REQUESTED** [review]
One top-level thought -- in Rust, the convention is to use `iter_*` for methods that return iterators. Here, this would mean that `get_itemcollections` would be renamed to `iter_itemcollections` (same with `iter_items`). Does that scan any better to you @philvarner? Just a thought as I was musing about our discussion on whether to include `all` in the method name.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,15 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 ## [Unreleased] - TBD
 
+### Deprecated
+
+- Item Search methods `get_all_items` and `get_all_items_as_dict` are now deprecated.
+  These have been deprecated because they have the potential to perform a large number
+  of requests to the server and instantiate a large number of objects in memory.
+  To a user, this is only visible as a large delay in the method call and/or the
+  exhaustion of all available memory. The iterator methods `get_items` or
+  `get_itemcollections` should be used instead. [#206](https://github.com/stac-utils/pystac-client/pull/206)
```
Comment (philvarner) [inline CHANGELOG.md:16]: also just added the PR links here.

File: CHANGELOG.md (hunk 1 above)
Comment (philvarner) [reply CHANGELOG.md]: I would be in favor of saying it's going to be removed in 0.5.

Comment (gadomski) [reply CHANGELOG.md]: 👍🏽, but I'd like to have @matthewhanson's ok as well if possible just b/c this is the first time we're deprecating in here (I think).

File: docs/usage.rst [hunk 4]
```diff
@@ -201,6 +198,35 @@ described in the
...
+    S2B_OPER_MSI_L2A_TL_SGS__20190101T200120_A009518_T18TXP_N02.11
+    MCD43A4.A2019010.h12v04.006.2019022234410
+    MCD43A4.A2019009.h12v04.006.2019022222645
+    MYD11A1.A2019002.h12v04.006.2019003174703
+    MYD11A1.A2019001.h12v04.006.2019002165238
+
+If you do not need the :class:`pystac.Item` instances, you can instead use
+:meth:`ItemSearch.get_items_as_dicts <pystac_client.ItemSearch.get_items_as_dicts>`
+to retrive dictionary representation of the items, without incurring the cost of
+creating the pydantic Item objects.
```
Comment (gadomski) [inline docs/usage.rst:218]:
```suggestion
creating the Item objects.
```

Sorry, I missed this other `pydantic` reference.

**Review by gadomski — CHANGES_REQUESTED** [review]

**Comment by philvarner** [comment]
> One top-level thought -- in Rust, the convention is to use iter_* for methods that return iterators. Here, this would mean that get_itemcollections would be renamed to iter_itemcollections (same with iter_items). Does that scan any better to you @philvarner? Just a thought as I was musing about our discussion on whether to include all in the method name.

Few thoughts. I like that convention, but it's not one used in Python, so I'd avoid it. Though, I also hate methods prefixed with "get" since that's the old (entirely unncecessary) OO convention for accessor methods. I also don't like using `get` when the underlying method has side-effects and isn't just returning a value. I usually prefer just to name it like `itemcollections()` or `items()` or `items_as_dicts()`. Since they are causing network requests, I might prefix them with `retrieve` to indicate that there's a lot going on under the hood.

File: CHANGELOG.md (hunk 1 above)
Comment (philvarner) [reply CHANGELOG.md]: also as a reminder, the semver convention is that minor versions prior to 1.0 are the same as major version after.

**Comment by philvarner** [comment]
> > Since they are causing network requests, I might prefix them with retrieve to indicate that there's a lot going on under the hood.
> 
> Agreed, though since they can be `GET` requests I guess `get_` isn't a terrible prefix either :-). Full circle.

Well, we default to POST, so... 🤣 

> I don't think I feel strongly enough to think a change is necessary now, but happy to support a change too with others' input.

oh, so apparently the get prefix was added by in v0.2.0, from the changelog:

```
- `Search.item_collections()` renamed to `Search.get_item_collections()`
- `Search.item_collections()` renamed to `Search.get_items()`
```

@matthewhanson any thoughts on this?



[1:42](https://element84.slack.com/archives/C03B8T0E6CC/p1654105321261179)
(though i think the second msg should say Search.items() renamed

File: CHANGELOG.md (hunk 1 above)
Comment (philvarner) [reply CHANGELOG.md]: I weasled it with "as early as v0.5.0"

**Review by gadomski — APPROVED** [review]

**Pull request merged**